File: LeetCode - Top Interview 150/74.py
# 74. Search a 2D Matrix
# https://leetcode.com/problems/search-a-2d-matrix/


from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row_left = 0
        row_right = len(matrix) - 1

        target_row = -1

        while row_left <= row_right:
            row_mid = (row_left + row_right) // 2

            if target < matrix[row_mid][0]:
                row_right = row_mid - 1
            elif target > matrix[row_mid][-1]:
                row_left = row_mid + 1
            else:
                target_row = row_mid

                break

        col_left = 0
        col_right = len(matrix[target_row]) - 1

        while col_left <= col_right:
            col_mid = (col_left + col_right) // 2

            if target == matrix[target_row][col_mid]:
                return True
            elif target < matrix[target_row][col_mid]:
                col_right = col_mid - 1
            else:
                col_left = col_mid + 1

        return False

File: LeetCode - Algorithm/Binary Search/74.js
// 74. Search a 2D Matrix
// https://leetcode.com/problems/search-a-2d-matrix/

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  const numRows = matrix.length;
  const numCols = matrix[0].length;

  const numberToCoordinate = (number) => {
    const quotient = Math.floor(number / numCols);
    const remainer = number % numCols;

    return [quotient, remainer];
  };

  let left = 0;
  let right = numRows * numCols - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);

    const [row, col] = numberToCoordinate(mid);
    const midNumber = matrix[row][col];

    if (midNumber === target) {
      return true;
    } else if (midNumber > target) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return false;
};
